[inductor] Adding a way to force fusion of int_mm with mul

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111125

Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

HDCharles committed:
WIP Adding int_mm_mul (int_mm fused with two muls)

Summary: unsure if there's a more general way to set it up rather than
forcing the order of teh mul's based on their shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Chillee commented:
I don't understand why we need a pattern for this? Why don't we just fix it so that it does prioritize this fusion?

cpuhrsch commented:
@Chillee - I agree with you. This kernel also uses block ptrs. It's a starting point. This isn't supposed to be the final version.

HDCharles committed:
Update on "[inductor] Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

HDCharles commented:
> I don't understand why we need a pattern for this? Why don't we just fix it so that it does prioritize this fusion?

yeah this would ideally be temporary until we fix it so it prioritizes the fusion, the reason we needed a pattern for this is mostly because i knew how to do that but couldn't figure out how to fix the fusion.

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -834,3 +834,45 @@ def repl(inp, mat1, mat2):
+    return (
+        len(getattr(match.args[2].meta.get("val"), "shape", [])) == 2
+        and getattr(match.args[2].meta.get("val"), "is_cuda", False)
+        and config.force_fuse_int_mm_with_mul
```
check this first

jansel commented on torch/_inductor/kernel/fused_int_mm_mul.py:
```diff
@@ -0,0 +1,97 @@
+
+log = logging.getLogger(__name__)
+
+fused_int_mm_mul_template = TritonTemplate(
```
You shouldn't need a new template for adding an epilogue to an existing template.  

See how addmm uses `epilogue_fn`:
https://github.com/pytorch/pytorch/blob/ad4472833ce5f40ca595492bd48f6735c69d6a3f/torch/_inductor/kernel/mm.py#L240

HDCharles replied:
thanks, I had been looking for an example of how to do that

jansel commented on torch/_inductor/kernel/fused_int_mm_mul.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def tuned_fused_int_mm_mul(mat1, mat2, mat3, out_dtype, *, layout=None):
+    out_dtype = mat3.get_dtype() if out_dtype is None else out_dtype
```
For the default dtype we should use `torch.promote_types`

HDCharles committed:
Update on "[inductor] Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

HDCharles committed:
Update on "[inductor] Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/kernel/mm.py:
```diff
@@ -286,3 +287,25 @@ def tuned_mixed_mm(mat1, mat2, mat2_dtype):
+
+def tuned_fused_int_mm_mul(mat1, mat2, mat3, out_dtype, *, layout=None):
+    out_dtype = (
+        torch.promote_types(mat3.get_dtype(), torch.int32)
```
```suggestion
        functools.reduce(torch.promote_types, [
             mat1.get_dtype(),
             mat2.get_dtype(),
             mat3.get_dtype()
        ])
```
I am assuming this will match the semantics of `aten._int_mm`, but you should check to make sure.

Will aten._int_mm promote to torch.int32 even if none of the inputs are int32?  What about int64?

HDCharles replied:
int_mm only does int8 int8 -> int32 matmuls so i don't think those others are possible.

with torch.promote_types(mat3.get_dtype(), torch.int32) i was trying to set the output type to be what you'd naturally get from an int32 * float16 so that it matches the non torch-compiled model. What is the benefit of including the mat1 and mat2 dtypes?

jansel commented on torch/_inductor/kernel/mm.py:
```diff
@@ -286,3 +287,25 @@ def tuned_mixed_mm(mat1, mat2, mat2_dtype):
     return autotune_select_algorithm("mixed_mm", choices, [mat1, mat2], layout)
+
+
+def tuned_fused_int_mm_mul(mat1, mat2, mat3, out_dtype, *, layout=None):
```
Add a comment explaining why there is no aten option here, and that this is only enabled by `config.force_fuse_int_mm_with_mul`

HDCharles committed:
Update on "[inductor] Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

HDCharles committed:
Update on "[inductor] Adding a way to force fusion of int_mm with mul"


Summary: When doing quantization int_mm -> mul or int_mm -> mul ->
to(dtype) is an extremely common op pattern which is currently not
handled well by inductor. Ideally, since the output of
int_mm has dtype int32 we'd prefer to only realize a smaller dtype like
bf16 or float16. Currently inductor doesn't have a way to force this, in
many cases the mul gets fused with a bunch of subsequent pointwise
ops from the dequant creating an increase in memory overhead and a general
slowdown compared to the fused version.

Theoretically with better control of/smarter inductor fusion, this could be something we get for free, at which point these changes can be removed.

Test Plan: python test/inductor/test_pattern_matcher.py -k
"int_mm_mul"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it fails on ROCm https://hud.pytorch.org/pytorch/pytorch/commit/f4297576e63e4110f6bdf2522ae6a5fb4c7f3816' -c nosignals

The error is `RuntimeError: _int_mm_out_cuda not compiled for this platform`.  ROCm is currently having some stability issue, so its tests doesn't block merge atm.  However, the issue here looks legit.

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it fails on ROCm https://hud.pytorch.org/pytorch/pytorch/commit/f4297576e63e4110f6bdf2522ae6a5fb4c7f3816' -c nosignal
